### ISSUE TITLE
Better guard against catastrophic cascade deletions

### ISSUE THREAD

**Issue by philipbelesky** [czlee/tabbycat#403]
There's been a few times when persons have unwittingly deleted results (or tournaments) unintentionally in the backend area. It would be good to have some hook into that page such that whenever a `TeamScore` or other critical piece of information would be deleted to very explicitly warn the user that results will be deleted — currently the list of cascade deletions is very long (and requires users to know what the various data types mean)

**Comment by czlee** [comment]
Just thinking—when you think about it, most deletions are "catastrophic" if accidental. What's our threshold? Should we limit to just those where the entire tournament or entire round is at stake?

Another option would be to have a message in the edit database area instructing users to back up their database first? We still don't have a backup guide in our documentation, actually…

**Comment by philipbelesky** [comment]
I think a good threshold would be any deletion that would cause a `TeamScore` / `SpeakerScore` to be deleted. So that applies to perhaps small accidents (deleting a debate) or big ones (deleting a round). In my mind those are the ones that are not easily undone and are the worst cases for accidental deletion.

**Comment by czlee** [comment]
Hmm, what about deleting an adjudicator allocation?

**Comment by philipbelesky** [comment]
Yes, I would consider that classic sufficiently dire *if* results existed for the debate which is having its aloocation removed.

**Comment by czlee** [comment]
Oh, that's a no, then (since the result is still a sufficient condition). Similarly, deleting a round is A-okay so long as no results, right?

I think we'd use a similar approach to what we did with our debate team warnings:
https://github.com/czlee/tabbycat/tree/develop/tabbycat/draw/templates/admin/draw/debateteam
https://docs.djangoproject.com/en/1.10/ref/contrib/admin/#overriding-admin-templates

**Comment by czlee** [comment]
So, how do we deal with changing adjudicators after a result's been entered? Currently it just goes ahead with the cascade deletions, which isn't catastrophic, but also isn't great. The tempting thing to do would be to protect `SpeakerScoreByAdj`'s `debate_adjudicator` field, which is probably a good idea, but without further work would just involve a bunch of 500s being thrown at the Ajax in the edit adjudicators screen. It's not actually clear to me even from the user's perspective what the intuitive behavior _should_ be.

**Comment by philipbelesky** [comment]
Hmm, it's probably rare and important enough that we could block it on the edit adjs screen if needed. I suspect most users would be unaware of the effects of changing the adjs, so we should probably lean on the side of overly cautioning them (and making them notice the issue) rather than silently fixing things?

**Comment by czlee** [comment]
I agree in principle. But there's still the question of what should happen if they decide to proceed. My inclination is just to warn them that it will delete all ballots for that debate, then do so if they confirm. If we want to do the swap-adjudicator thing then the front-end will have to hold transmission until the swap is complete (i.e. both remove and add).

**Comment by philipbelesky** [comment]
Is this likely to make it into Iberian Lynx, or should it be bumped?

**Comment by czlee** [comment]
Do we have a concrete picture of what this looks like yet? Re-reading the thread I'm finding it hard to convince myself that this should be a priority.

**Comment by philipbelesky** [comment]
I'm guessing it would in an ideal world look like an intermediate step or more aggressive set of warnings when doing the delete operation in the Edit Database area. But I can't think of recent issue that would require this or lead it to be a priority.
